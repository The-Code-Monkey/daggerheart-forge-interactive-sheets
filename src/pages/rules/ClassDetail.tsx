import { useParams, Link } from "react-router-dom";
import {
  Card,
  CardContent,
  CardDescription,
  CardHeader,
  CardTitle,
} from "@/components/ui/card";
import { Badge } from "@/components/ui/badge";
import { Button } from "@/components/ui/button";
import { ArrowLeft } from "lucide-react";
import { JSX, useEffect, useState } from "react";
import { Class } from "@/lib/types";
import { Tabs, TabsContent, TabsList, TabsTrigger } from "@/components/ui/tabs";
import { getSingleClassBySlugWithDomains } from "@/integrations/supabase/helpers";

const ClassDetail = (): JSX.Element => {
  const { className } = useParams();
  const [classData, setClassData] = useState<Class | null>(null);
  const [loading, setLoading] = useState(true);

  const fetchClass = async () => {
    const data = await getSingleClassBySlugWithDomains(String(className));

    setClassData(data);
    setLoading(false);
  };

  useEffect(() => {
    void fetchClass();
  }, [className]);

  if (loading) {
    return (
      <div className="min-h-screen bg-gradient-to-b from-purple-900 via-purple-800 to-slate-900 py-8 px-4">
        <div className="max-w-4xl mx-auto text-center">
          <h1 className="text-4xl font-bold text-white mb-4">
            Loading Class...
          </h1>
        </div>
      </div>
    );
  }

  if (!classData) {
    return (
      <div className="min-h-screen bg-gradient-to-b from-purple-900 via-purple-800 to-slate-900 py-8 px-4">
        <div className="max-w-4xl mx-auto text-center">
          <h1 className="text-4xl font-bold text-white mb-4">
            Class Not Found
          </h1>
          <Link to="/game-rules">
            <Button className="bg-purple-600 hover:bg-purple-700">
              <ArrowLeft className="w-4 h-4 mr-2" />
              Back to Game Rules
            </Button>
          </Link>
        </div>
      </div>
    );
  }

  const subclasses = Object.keys(classData.additional.subclasses ?? {});

  return (
    <div className="min-h-screen bg-gradient-to-b from-purple-900 via-purple-800 to-slate-900 py-8 px-4">
      <div className="max-w-4xl mx-auto">
        {/* Header */}
        <div className="mb-8">
          <Link to="/game-rules">
            <Button
              variant="ghost"
              className="text-purple-200 hover:text-white mb-4"
            >
              <ArrowLeft className="w-4 h-4 mr-2" />
              Back to Game Rules
            </Button>
          </Link>

          <div className="flex items-center gap-4 mb-4">
            <div className="w-16 h-16 bg-gradient-to-br from-yellow-400 to-orange-500 rounded-lg flex items-center justify-center">
              {/* {classData.icon} */}
            </div>
            <div>
              <h1 className="text-4xl font-bold text-white">
                {classData.name}
              </h1>
              <p className="text-xl text-purple-200">{classData.description}</p>
            </div>
          </div>
        </div>

        {/* Class Stats */}
        <div className="grid grid-cols-1 md:grid-cols-2 gap-6 mb-8">
          <Card className="bg-gradient-to-br from-slate-800/80 to-slate-900/80 border-purple-500/30 backdrop-blur-sm">
            <CardHeader>
              <CardTitle className="text-white">Hit Points</CardTitle>
            </CardHeader>
            <CardContent>
              <p className="text-purple-200">{classData.base_hp}</p>
            </CardContent>
          </Card>

          <Card className="bg-gradient-to-br from-slate-800/80 to-slate-900/80 border-purple-500/30 backdrop-blur-sm">
            <CardHeader>
              <CardTitle className="text-white">Evasion</CardTitle>
            </CardHeader>
            <CardContent>
              <p className="text-purple-200">{classData.base_evasion}</p>
            </CardContent>
          </Card>

          <Card className="bg-gradient-to-br from-slate-800/80 to-slate-900/80 border-purple-500/30 backdrop-blur-sm">
            <CardHeader>
              <CardTitle className="text-white">Available Domains</CardTitle>
            </CardHeader>
            <CardContent>
              <div className="flex flex-wrap gap-1">
                {classData.domains.map((domain) => (
                  <Badge
                    key={domain.id}
                    variant="outline"
                    className="border-yellow-500/50 text-yellow-300 text-xs"
                  >
                    {domain.name}
                  </Badge>
                ))}
              </div>
            </CardContent>
          </Card>

          <Card className="bg-gradient-to-br from-slate-800/80 to-slate-900/80 border-purple-500/30 backdrop-blur-sm">
            <CardHeader>
              <CardTitle className="text-white">Class Items</CardTitle>
            </CardHeader>
            <CardContent>
              <p className="text-purple-200">{classData.class_items}</p>
            </CardContent>
          </Card>
        </div>

        {/* Class Features */}
        <Card className="bg-gradient-to-br from-slate-800/80 to-slate-900/80 border-purple-500/30 backdrop-blur-sm col-span-2">
          <CardHeader>
            <CardTitle className="text-white">Class Features</CardTitle>
            <CardDescription className="text-purple-200">
              Special abilities that define the {classData.name} class
            </CardDescription>
          </CardHeader>
          <CardContent>
            <div className="space-y-4">
              {classData.features.map((feature, index) => (
                <div key={`feature-${String(index)}`}>
                  <h3 className="text-lg font-semibold text-white mb-2">
                    {feature.name}
                  </h3>
                  <p className="text-purple-200">{feature.description}</p>
                  {feature.list && (
                    <ul className="list-disc list-inside">
                      {feature.list.map((item, itemIndex) => (
                        <li
                          key={`feature-list-${String(itemIndex)}`}
                          className="text-purple-200"
                        >
                          {item}
                        </li>
                      ))}
                    </ul>
                  )}
                </div>
              ))}
            </div>
          </CardContent>
        </Card>

        <Tabs defaultValue={subclasses[0]} className="w-full mt-6">
          <TabsList className="w-full justify-evenly bg-purple-800/50 mb-8">
            {subclasses.map((key) => {
              const subclass = classData.additional.subclasses![key];

              return (
                <TabsTrigger
                  key={key}
                  value={key}
                  className="text-white data-[state=active]:bg-purple-600 capitalize"
                >
                  {subclass.name}
                </TabsTrigger>
              );
            })}
          </TabsList>

          {subclasses.map((key) => {
            const subclass = classData.additional.subclasses![key];

            return (
              <TabsContent value={key} className="space-y-6">
                <Card className="bg-gradient-to-br from-slate-800/80 to-slate-900/80 border-purple-500/30 backdrop-blur-sm flex flex-col">
                  <CardHeader>
                    <CardTitle className="text-white">
                      {subclass.name}
                    </CardTitle>
                  </CardHeader>
                  <CardContent>
                    <div className="space-y-4">
                      <div>
                        <h3 className="text-xl font-bold text-white mb-2 capitalize">
                          Foundation Features
                        </h3>
                        {subclass.features?.foundation.map(
                          (feature, featureIndex) => (
                            <div key={featureIndex} className="text-white">
                              <p className="inline-block">
                                <span className="font-bold">
                                  {feature.name}:
                                </span>{" "}
                                {feature.description}
                              </p>
                              {feature.list && (
                                <ul className="list-disc pl-4 my-2">
                                  {feature.list.map(
                                    (listItem, listItemIndex) => (
                                      <li
                                        key={listItemIndex}
                                        className="text-white"
                                      >
                                        {listItem}
                                      </li>
                                    )
                                  )}
                                </ul>
                              )}
                            </div>
                          )
                        )}
                      </div>
                      <hr />
                      <div>
                        <h3 className="text-lg font-semibold text-white mb-2 capitalize">
                          Specialization Features
                        </h3>
                        {subclass.features?.specialization.map(
                          (feature, featureIndex) => (
                            <div key={featureIndex} className="text-white">
                              <p className="inline-block">
                                <span className="font-bold">
                                  {feature.name}:
                                </span>{" "}
                                {feature.description}
                              </p>
                            </div>
                          )
                        )}
                      </div>
                      <hr />
                      <div>
                        <h3 className="text-lg font-semibold text-white mb-2 capitalize">
                          Mastery Features
                        </h3>
                        {subclass.features?.mastery.map(
                          (feature, featureIndex) => (
                            <div key={featureIndex} className="text-white">
                              <p className="inline-block">
                                <span className="font-bold">
                                  {feature.name}:
                                </span>{" "}
                                {feature.description}
                              </p>
                            </div>
                          )
                        )}
                      </div>
                    </div>
                  </CardContent>
                </Card>
              </TabsContent>
            );
          })}
        </Tabs>
      </div>
    </div>
  );
};

export default ClassDetail;
