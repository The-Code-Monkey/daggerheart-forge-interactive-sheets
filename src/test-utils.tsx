import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

// Mock AuthContext
const mockAuthContext = {
  user: null,
  loading: false,
  signIn: jest.fn(),
  signOut: jest.fn(),
  signUp: jest.fn(),
};

const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  return <div data-testid="auth-provider">{children}</div>;
};

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <BrowserRouter>
      <AuthProvider>{children}</AuthProvider>
    </BrowserRouter>
  );
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

// Export everything from testing-library/react
export * from "@testing-library/react";
export { customRender as render };

// Mock useAuth hook
export const mockUseAuth = () => mockAuthContext;
