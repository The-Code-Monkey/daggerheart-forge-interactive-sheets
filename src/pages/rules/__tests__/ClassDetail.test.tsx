import { render, act } from "../../../test-utils";
import ClassDetail from "../ClassDetail";

// Mock react-router-dom
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({ className: "guardian" }),
  Link: ({ children, to }: { children: React.ReactNode; to: string }) => (
    <a href={to}>{children}</a>
  ),
}));

// Mock supabase helpers
jest.mock("@/integrations/supabase/helpers", () => ({
  getSingleClassBySlugWithDomains: jest.fn().mockResolvedValue({
    id: 1,
    name: "Guardian",
    slug: "guardian",
    description: "Protector and defender",
    base_hp: 10,
    base_evasion: 10,
    class_items: "Shield, Sword",
    features: [
      {
        name: "Test Feature",
        description: "Test description",
      },
    ],
    domains: [
      {
        id: 1,
        name: "Protection",
      },
    ],
    additional: {
      subclasses: {
        armor: {
          name: "Armor",
          features: {
            foundation: [
              {
                name: "Foundation Feature",
                description: "Foundation description",
              },
            ],
            specialization: [
              {
                name: "Specialization Feature",
                description: "Specialization description",
              },
            ],
            mastery: [
              {
                name: "Mastery Feature",
                description: "Mastery description",
              },
            ],
          },
        },
      },
    },
  }),
}));

describe("ClassDetail Page", () => {
  it("renders correctly", async () => {
    let container;
    await act(async () => {
      const { container: cont } = render(<ClassDetail />);
      container = cont;
    })
    expect(container.firstChild).toMatchSnapshot();
  });
});
