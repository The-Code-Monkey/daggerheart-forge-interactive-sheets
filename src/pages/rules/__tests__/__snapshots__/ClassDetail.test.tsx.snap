// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassDetail Page renders correctly 1`] = `
<div
  data-testid="auth-provider"
>
  <div
    class="min-h-screen bg-nebula py-8 px-4"
  >
    <div
      class="max-w-4xl mx-auto"
    >
      <div
        class="mb-8"
      >
        <a
          href="/"
        >
          <button
            class="inline-flex items-center hover:cursor-pointer justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors duration-300 ease-in-out focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary hover:bg-primary/90 h-10 px-4 py-2 text-purple-200 hover:text-white mb-4"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-arrow-left w-4 h-4 mr-2"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m12 19-7-7 7-7"
              />
              <path
                d="M19 12H5"
              />
            </svg>
            Go Back
          </button>
        </a>
        <div
          class="flex items-center gap-4 mb-4"
        >
          <div
            class="w-16 h-16 bg-linear-to-br from-yellow-400 to-orange-500 rounded-lg flex items-center justify-center"
          />
          <div>
            <h1
              class="text-4xl font-bold text-white"
            >
              Guardian
            </h1>
            <p
              class="text-xl text-purple-200"
            >
              Protector and defender
            </p>
          </div>
        </div>
      </div>
      <div
        class="grid grid-cols-1 md:grid-cols-2 gap-6 mb-8"
      >
        <div
          class="rounded-lg border text-card-foreground shadow-sm bg-gradient-to-br bg-linear-to-br from-slate-800/80 to-slate-900/80 border-purple-500/30 backdrop-blur-xs"
        >
          <div
            class="flex flex-col space-y-1.5 p-6"
          >
            <h3
              class="text-2xl font-semibold leading-none tracking-tight text-white"
            >
              Hit Points
            </h3>
          </div>
          <div
            class="p-6 pt-0"
          >
            <p
              class="text-purple-200"
            >
              10
            </p>
          </div>
        </div>
        <div
          class="rounded-lg border text-card-foreground shadow-sm bg-gradient-to-br bg-linear-to-br from-slate-800/80 to-slate-900/80 border-purple-500/30 backdrop-blur-xs"
        >
          <div
            class="flex flex-col space-y-1.5 p-6"
          >
            <h3
              class="text-2xl font-semibold leading-none tracking-tight text-white"
            >
              Evasion
            </h3>
          </div>
          <div
            class="p-6 pt-0"
          >
            <p
              class="text-purple-200"
            >
              10
            </p>
          </div>
        </div>
        <div
          class="rounded-lg border text-card-foreground shadow-sm bg-gradient-to-br bg-linear-to-br from-slate-800/80 to-slate-900/80 border-purple-500/30 backdrop-blur-xs"
        >
          <div
            class="flex flex-col space-y-1.5 p-6"
          >
            <h3
              class="text-2xl font-semibold leading-none tracking-tight text-white"
            >
              Available Domains
            </h3>
          </div>
          <div
            class="p-6 pt-0"
          >
            <div
              class="flex flex-wrap gap-1"
            >
              <div
                class="inline-flex items-center rounded-full border px-2.5 py-0.5 font-semibold transition-colors focus:outline-hidden focus:ring-2 focus:ring-ring focus:ring-offset-2 border-yellow-500/50 text-yellow-300 text-xs"
              >
                Protection
              </div>
            </div>
          </div>
        </div>
        <div
          class="rounded-lg border text-card-foreground shadow-sm bg-gradient-to-br bg-linear-to-br from-slate-800/80 to-slate-900/80 border-purple-500/30 backdrop-blur-xs"
        >
          <div
            class="flex flex-col space-y-1.5 p-6"
          >
            <h3
              class="text-2xl font-semibold leading-none tracking-tight text-white"
            >
              Class Items
            </h3>
          </div>
          <div
            class="p-6 pt-0"
          >
            <p
              class="text-purple-200"
            >
              Shield, Sword
            </p>
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border text-card-foreground shadow-sm bg-gradient-to-br bg-linear-to-br from-slate-800/80 to-slate-900/80 border-purple-500/30 backdrop-blur-xs col-span-2"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
        >
          <h3
            class="text-2xl font-semibold leading-none tracking-tight text-white"
          >
            Class Features
          </h3>
          <p
            class="text-sm text-purple-200"
          >
            Special abilities that define the 
            Guardian
             class
          </p>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="space-y-4"
          >
            <div>
              <h3
                class="text-lg font-semibold text-white mb-2"
              >
                Test Feature
              </h3>
              <p
                class="text-purple-200"
              >
                Test description
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="w-full mt-6"
        data-orientation="horizontal"
        dir="ltr"
      >
        <div
          aria-orientation="horizontal"
          class="inline-flex items-center rounded-md p-1 text-muted-foreground bg-[rebeccapurple] gap-2 w-full justify-evenly mb-8"
          data-orientation="horizontal"
          role="tablist"
          style="outline: none;"
          tabindex="0"
        >
          <button
            aria-controls="radix-«r0»-content-Armor"
            aria-selected="true"
            class="inline-flex items-center justify-center whitespace-nowrap rounded-sm px-3 py-1.5 text-sm font-medium ring-offset-background transition-all focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:cursor-pointer data-[state=active]:shadow-xs data-[state=active]:text-white hover:bg-purple-600/75 duration-300 ease-in-out text-white data-[state=active]:bg-purple-600 flex-1 capitalize"
            data-orientation="horizontal"
            data-radix-collection-item=""
            data-state="active"
            id="radix-«r0»-trigger-Armor"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Armor
             
          </button>
        </div>
        <div
          aria-labelledby="radix-«r0»-trigger-Armor"
          class="mt-2 ring-offset-background focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 space-y-6"
          data-orientation="horizontal"
          data-state="active"
          id="radix-«r0»-content-Armor"
          role="tabpanel"
          style="animation-duration: 0s;"
          tabindex="0"
        >
          <div
            class="rounded-lg border text-card-foreground shadow-sm bg-gradient-to-br bg-linear-to-br from-slate-800/80 to-slate-900/80 border-purple-500/30 backdrop-blur-xs flex flex-col"
          >
            <div
              class="flex flex-col space-y-1.5 p-6"
            >
              <h3
                class="text-2xl font-semibold leading-none tracking-tight text-white"
              >
                Armor
              </h3>
            </div>
            <div
              class="p-6 pt-0"
            >
              <div
                class="space-y-4"
              >
                <div>
                  <h3
                    class="text-xl font-bold text-white mb-2 capitalize"
                  >
                    Foundation Features
                  </h3>
                  <div
                    class="text-white"
                  >
                    <p
                      class="inline-block"
                    >
                      <span
                        class="font-bold"
                      >
                        Foundation Feature
                        :
                      </span>
                       
                      Foundation description
                    </p>
                  </div>
                </div>
                <hr />
                <div>
                  <h3
                    class="text-lg font-semibold text-white mb-2 capitalize"
                  >
                    Specialization Features
                  </h3>
                  <div
                    class="text-white"
                  >
                    <p
                      class="inline-block"
                    >
                      <span
                        class="font-bold"
                      >
                        Specialization Feature
                        :
                      </span>
                       
                      Specialization description
                    </p>
                  </div>
                </div>
                <hr />
                <div>
                  <h3
                    class="text-lg font-semibold text-white mb-2 capitalize"
                  >
                    Mastery Features
                  </h3>
                  <div
                    class="text-white"
                  >
                    <p
                      class="inline-block"
                    >
                      <span
                        class="font-bold"
                      >
                        Mastery Feature
                        :
                      </span>
                       
                      Mastery description
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
